app/grandchallenge/core/templatetags/bleach.py

```diff
@@ -6,18 +8,45 @@
 from markdown import markdown as render_markdown
 from markdown.extensions.toc import TocExtension
 
-from grandchallenge.core.utils.markdown import LinkBlankTargetExtension
+from grandchallenge.core.utils.markdown import (
+    EmbedYoutubeExtension,
+    LinkBlankTargetExtension,
+)
 
 register = template.Library()
 
 
+def allowed_iframe(tag, name, value):
+    """Returns true if the iframe tag is allowed, false otherwise"""
+    if tag != "iframe":
+        return False
+    if name in settings.BLEACH_ALLOWED_FRAME_ATTRIBUTES:
+        return True
+    if name == "src":
```

Comment:
Is there a way to only allow iframes with `sandbox`?

Reply:
Yes, but not straightforward. nh3 doesn't seem to support it either. We could introduce a custom markdown extension for that. But that runs before the cleaning. 

It also has minimum benefit: we'll allow for 'weak' sandboxes so someone could still use a `sandbox="allow-x allow-y"` and have a not-effective sandbox. We could introduce sub-attribute checks.... but that might be going overboard.

Reply:
I'm wondering if we're going about this the wrong way because we don't have much control here. Maybe we should first convert the markdown to html and end up with a safestring. Then, we do our own post-processing on that and insert our own html, then no messing about with bleach. We used to have such a thing that worked with `{% tag argument %}` but it could be easily changed for `[tag argument]`:

https://github.com/comic/grand-challenge.org/blob/c2737876c91e15626f2e6908af741279e22d588a/app/grandchallenge/pages/substitutions.py

https://github.com/comic/grand-challenge.org/blob/c2737876c91e15626f2e6908af741279e22d588a/app/tests/pages_tests/test_substitutions.py

Reply:
I thought your suggestion was a very good one: why bother with bleaching and clean content when we actually generate the HTML ourselves! As such I have re-introduced the `Substitution` class but named it `TagSubstitution` and updated a bit to work with Callables.

Extending with new custom tags should now be really, _really_ easy.